### PR TITLE
vendor newest libelf

### PR DESCRIPTION
**Pull request by namtsui**
I noticed that libelf is licensed under Australian Public Licence B (OZPLB). I propose moving it to a newer version because it seems newer libelf has moved to BSD 2-Clause.

source:
https://github.com/seL4/util_libs/tree/2d5b59c99a93507266c10efe50760f25b7d73728/libelf

license question:
https://github.com/seL4/util_libs/issues/150

I adapted reios_elf.cpp by styling it after the new core/deps/libelf/src/elf.c elf_loadFile().

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -198,7 +198,7 @@ if(NOT "${SENTRY_UPLOAD_URL}" STREQUAL "")
 	target_compile_definitions(${PROJECT_NAME} PRIVATE SENTRY_UPLOAD="${SENTRY_UPLOAD_URL}")
 endif()
 
-target_include_directories(${PROJECT_NAME} PRIVATE core core/deps core/deps/stb core/khronos core/deps/json)
+target_include_directories(${PROJECT_NAME} PRIVATE core core/deps core/deps/stb core/khronos core/deps/json core/deps/libelf/include)
```
Comment (scribam) [inline CMakeLists.txt:201]: Is it really necessary? This directory is already included with `target_include_directories(elf PUBLIC include)` from `core/deps/libelf/CMakeLists.txt`

**Comment by flyinghead** [comment]
Did you test some dreamcast .elf files with these changes?

**Comment by namtsui** [comment]
I've only booted up flycast games like normal like with Marvel vs Capcom 2 chd file and assumed once it hit reios/reios.cpp that it would test it. I'm not sure how to test it.

```
$ flycast
00:00:824 sdl/sdl.cpp:603 N[RENDERER]: Monitor refresh rate: 60 Hz (1920 x 1080)
00:00:830 rend/gui.cpp:273 N[RENDERER]: Screen DPI is 91, size 958 x 1052. Scaling by 1.00
00:00:837 rend/gles/gles.cpp:563 N[RENDERER]: OpenGL version 4.5
00:13:616 hw/mem/_vmem.cpp:473 N[VMEM]: Info: nvmem is enabled
00:13:659 hw/mem/_vmem.cpp:512 N[VMEM]: BASE 0xb155f270000 RAM(16 MB) 0xb156b270000 VRAM64(8 MB) 0xb1563270000 ARAM(2 MB) 0xb157f270000
00:13:671 hw/mem/_vmem.cpp:473 N[VMEM]: Info: nvmem is enabled
00:13:680 hw/mem/_vmem.cpp:512 N[VMEM]: BASE 0xb155f270000 RAM(16 MB) 0xb156b270000 VRAM64(8 MB) 0xb1563270000 ARAM(2 MB) 0xb157f270000
00:13:694 emulator.cpp:52 N[BOOT]: Game ID is [T1212N]
00:13:694 emulator.cpp:150 N[BOOT]: Japan region not supported. Using USA instead
00:13:696 emulator.cpp:518 N[BOOT]: Did not load BIOS, using reios
00:13:698 emulator.cpp:52 N[BOOT]: Game ID is [T1212N]
00:13:716 reios/reios.cpp:615 N[REIOS]: -----------------
00:13:716 reios/reios.cpp:616 N[REIOS]: REIOS: Booting up
00:13:716 reios/reios.cpp:617 N[REIOS]: -----------------
```

**Comment by namtsui** [comment]
I've only confirmed that booting up works for cvs2 (naomi), mvc2 (naomi), mvc2 usa dreamcast, phantasy star online and cannon spike. I'll playtest through like a game in each and report back.

**Comment by namtsui** [comment]
Yeah I was able to playtest these games.

**Comment by flyinghead** [comment]
Flycast can directly load and run dreamcast .elf flles (prototypes, homebrew, tests, etc.) This is what libelf is used for.

I'll do some tests before merging then.

**Comment by flyinghead** [comment]
Works fine

**Comment by sjnewbury** [comment]
Unless the intention is to have the flycast binary eventually link to the system libelf, this probably needs to be statically linked. 

```
ldd /usr/bin/flycast
	linux-vdso.so.1 (0x00007ffff7fc8000)
	...
	libelf.so => /usr/lib64/libelf.so (0x00007ffff7a9d000)
	...

```
/usr/lib64/libelf.so links to libelf-0.188.so from elfutils.

**Comment by flyinghead** [comment]
I don't see any dependency on libelf.so on my locally built flycast (ubuntu 20). I also checked the latest AppImage master build and no libelf.so dep either.
What distrib are you on and how did you get this flycast build?

**Comment by jeandestouches** [comment]
Hi,
Since this change, I'm unable to run flycast on my gentoo system. (2.1 release runs fine)

flycast: symbol lookup error: flycast: undefined symbol: elf_checkFile

build log: https://gist.github.com/jeandestouches/9d0b00a095615036a500fc975e5208d8

**Comment by flyinghead** [comment]
it looks like gentoo has cmake's `BUILD_SHARED_LIBS` enabled by default, which causes this error. 
Try building with `cmake -DBUILD_SHARED_LIBS=OFF`

**Comment by jeandestouches** [comment]
@flyinghead: Cool! Rebuilt using https://github.com/flyinghead/flycast/commit/9b8fd39d12f0add8db9512a30eda978400d1580c
Compile and runs fine now, thanks!

**Pull request merged**